clippy_lints/src/option_if_let_else.rs

```diff
@@ -141,7 +153,9 @@ fn detect_option_if_let_else<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) ->
             // with the borrow checker. Currently only `<local>[.<field>]*` is checked for.
             if as_ref || as_mut {
                 let e = peel_hir_expr_while(cond_expr, |e| match e.kind {
-                    ExprKind::Field(e, _) | ExprKind::AddrOf(_, _, e) => Some(e),
+                    ExprKind::Field(e, _)
+                    | ExprKind::AddrOf(_, _, e)
+                    | ExprKind::MethodCall(_, _, [e, ..], _) => Some(e),
```

Comment:
This seems incomplete. In the change higher up you look at the first `MethodCall` but here you can peel any number of method calls. Also, a function may take a mutable reference as input without necessarily returning a value that borrows the input.

This is a tricky problem and we have a lot of bugs like this. I think what we really need is a util like:

```rust
fn locals_borrowed_by_expr(&Expr) -> HirIdMap<Mutability>
```

and I think it should share some logic used by `ExprUseVisitor` but that logic is currently not public to Clippy.

Reply:
The more I look at this, the more I feel that this is actually hacky.

There is a recent issue that got opened for this specifically #8346.

Reply:
@dswij Are you okay with closing this PR, or do you want to give this another try?